tests/unit/test_solvers.py

```diff
@@ -0,0 +1,376 @@
+            lambda x: x.coreset_size, solver, len(dataset) + 1
+        )
+        with pytest.raises(ValueError, match=re.escape(self.OVERSIZE_MSG)):
+            modified_solver.reduce(dataset)
+
+
+class TestKernelHerding(RefinementSolverTest, ExplicitSizeSolverTest):
+    """Test cases for :class:`coreax.solvers.coresubset.KernelHerding`."""
+
+    @override
+    @pytest.fixture
+    def solver_factory(self) -> jtu.Partial:
+        kernel = PCIMQKernel()
+        coreset_size = self.shape[0] // 10
+        return jtu.Partial(KernelHerding, coreset_size=coreset_size, kernel=kernel)
+
+    @override
+    @pytest.fixture(params=["random", "analytic"])
+    def reduce_problem(
+        self, request, solver_factory: Union[type[Solver], jtu.Partial]
+    ) -> _ReduceProblem:
+        if request.param == "random":
+            dataset = jr.uniform(self.random_key, self.shape)
+            solver = solver_factory()
+            expected_coreset = None
+        elif request.param == "analytic":
+            dataset = jnp.array([[0, 0], [1, 1], [2, 2]])
+            # Set the kernel such that a simple analytic solution exists.
+            kernel_matrix = jnp.asarray([[1, 1, 1], [1, 1, 1], [0.5, 0.5, 0.5]])
+            kernel = MagicMock(Kernel)
+            kernel.compute = lambda _, y: kernel_matrix[:, y[0]]
+            kernel.gramian_row_mean.return_value = jnp.asarray([0.6, 0.75, 0.55])
+            solver = KernelHerding(coreset_size=2, kernel=kernel, unique=True)
+            expected_coreset = Coresubset(jnp.array([1, 2]), Data(dataset))
+        else:
+            raise ValueError("Invalid fixture parametrization")
+        return _ReduceProblem(Data(dataset), solver, expected_coreset)
+
+    def test_herding_state(self, reduce_problem: _ReduceProblem) -> None:
+        """Check that the cached herding state is as expected."""
+        dataset, solver, _ = reduce_problem
+        solver = cast(KernelHerding, solver)
+        _, state = solver.reduce(dataset)
+        expected_state = HerdingState(solver.kernel.gramian_row_mean(dataset))
+        assert eqx.tree_equal(state, expected_state)
+
+
+class TestRandomSample(ExplicitSizeSolverTest):
+    """Test cases for :class:`coreax.solvers.coresubset.RandomSample`."""
+
+    @override
+    def check_solution_invariants(
+        self, coreset: Coreset, problem: Union[_RefineProblem, _ReduceProblem]
+    ) -> None:
+        super().check_solution_invariants(coreset, problem)
+        solver = cast(RandomSample, problem.solver)
+        if solver.unique:
+            _, counts = jnp.unique(coreset.nodes.data, return_counts=True)
+            assert max(counts) <= 1
+
+    @override
+    @pytest.fixture
+    def solver_factory(self) -> jtu.Partial:
+        coreset_size = self.shape[0] // 10
+        key = jr.fold_in(self.random_key, self.shape[0])
+        return jtu.Partial(RandomSample, coreset_size=coreset_size, random_key=key)
+
+
+class TestRPCholesky(ExplicitSizeSolverTest):
```

Comment:
Just checking reasoning behind RPCholesky and SteinThinning not having their own overridden reduce_problem like KernelThinning. Guessing we do not have simple analytical examples for the former two algs?

Reply:
Yes, this is the reason for #628